src/charm.py

```diff
@@ -1,104 +1,377 @@
+            return False
+
+        unit_http_secrets = self.secrets.get_object(Scope.UNIT, CertType.UNIT_HTTP.val)
+        if unit_http_secrets is None or unit_http_secrets.get("cert") is None:
+            return False
+
+        return True
+
+    def _start_opensearch(self) -> bool:
+        """Start OpenSearch if all resources configured."""
+        # if there are any missing system requirements leave
+        missing_sys_reqs = self.opensearch.missing_sys_requirements()
+        if len(missing_sys_reqs) > 0:
+            self.unit.status = BlockedStatus(" - ".join(missing_sys_reqs))
+            return False
+
+        try:
+            self.unit.status = BlockedStatus("Waiting for OpenSearch to start...")
+            self.opensearch.start()
+            self.unit.status = ActiveStatus()
+
+            return True
+        except OpenSearchStartError:
+            return False
+
+    def _store_tls_resources(
+        self, cert_type: CertType, secrets: Dict[str, any], override_admin: bool = True
+    ):
+        """Write certificates and keys on disk."""
+        certs_dir = self.opensearch.paths.certs
+
+        self.opensearch.write_file(
+            f"{certs_dir}/{cert_type.val}.key",
+            to_pkcs8(secrets["key"], secrets.get("key-password")),
+        )
+        self.opensearch.write_file(f"{certs_dir}/{cert_type.val}.cert", secrets["cert"])
+        self.opensearch.write_file(f"{certs_dir}/root-ca.cert", secrets["ca"], override=False)
+
+        if cert_type == CertType.APP_ADMIN:
+            self.opensearch.write_file(
+                f"{certs_dir}/chain.pem",
+                "\n".join(secrets["chain"][::-1]),
+                override=override_admin,
+            )
+
+    def _initialize_admin_user(self):
+        """Change default password of Admin user."""
+        hashed_pwd, pwd = generate_hashed_password()
+        self.secrets.put(Scope.APP, "admin_password", pwd)
+
+        self.opensearch.config.put(
+            "opensearch-security/internal_users.yml",
+            "admin",
+            {
+                "hash": hashed_pwd,
+                "reserved": True,  # this protects this resource from being updated on the dashboard or rest api
+                "backend_roles": ["admin"],
+                "description": "Admin user",
+            },
+        )
+
+    def _should_init_security_index(self) -> bool:
+        """Evaluate whether we should set the security index."""
+        return (
+            self.unit.is_leader()
+            and self.app_peers_data.get("security_index_initialised") is None
+            and self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val) is not None
+        )
+
+    def _initialize_security_index(self, admin_secrets: Dict[str, any]):
+        """Run the security_admin script, it creates and initializes the opendistro_security index.
+
+        IMPORTANT: must only run once per cluster, otherwise the index gets overrode
```

Comment:
Is there a way to enforce this?

Reply:
Yes, we test whether the unit is the leader unit and whether the `security_index_initialised` flag on the app_data is set or not